File: admin_python/NVH/hdf_reader.py

import struct
from NVH.channel_data_model import ChannelDataModel


class HdfReader:
    baseBlockFrequency = 1000; # assume 1kHz, should be parsed from delta value
    
    # Constructor.
    def __init__(self, path):
        self.hdfFilePath = path
        self.dataLength = 0
        self.startPoint = 65536 # first candidate
        self.numberOfChannel = 0
        self.nbrBlock = 0
        self.channels = []


    def parseSync(self):
        self._parseHeader()

        #find data start point
        self.dataLength = self._getDataLength()

        # parse Raw data to Channel data model
        self._getChannelData()

        #return channels;

  
    def _parseHeader(self):
        f = open(self.hdfFilePath, 'rb')
        lines = f.readlines(self.startPoint)
        channelLines = []


        for i, lineBytes in enumerate(lines):
            line = str(lineBytes)
            if "start of data" in line :
                self.startPoint = int(line.split(":")[-1].strip().replace("\\r\\n'", ""))
                print("startPoint: ", self.startPoint)
            
            if "ch order" in line:
                chOrder = (line.split(":")[-1].strip().replace("\\r\\n'", ""))
                print("Channel Order: ",chOrder)

            if "scan mode" in line:
                mode = line.split(":")[-1].strip().replace("\\r\\n'", "")
                if mode != "synchronised multiple":
                    assert False

            if "nbr of scans" in line:
                self.nbrBlock = int(line.split(":")[-1].strip().replace("\\r\\n'", ""))
                print("Number of Block: ", self.nbrBlock)
            
            if "distribution func" in line:
                channelLines = lines[i:]
                break
        
        f.close()
        self._parseChannels(channelLines, chOrder)
    
    def _parseChannels(self, channelLinesBytes, chOrder):
        chOrders = chOrder.split(",")

        index = 0 
        for ch in chOrders:
            name = ""
            unit = ""
            frequency = 0
            while True:
                line = str(channelLinesBytes[index])
                index = index + 1
                if "name str" in line:
                    name = (line.split("name str:")[-1].strip().replace("\\r\\n'", ""))
                if "physical unit" in line:
                    unit = (line.split(":")[-1].strip().replace("\\r\\n'", ""))
                    break

            if "*" in ch:
                frequency = int(ch.split("*")[0])
            else:
                frequency = 1

            frequency *= self.baseBlockFrequency # assume block is based on 1kHz.

            self.channels.append(ChannelDataModel(name, unit, frequency))
            print("name:", name, "unit:", unit, "frequency:", frequency)

    def _getDataLength(self): 
        dataLength = 0
        f = open(self.hdfFilePath, 'rb')
        f.seek(self.startPoint -8192)
        dummy = f.read(8192)
        dummyString = str(dummy)
        if "data1" in dummyString: 
            dataLength = int(dummyString.split("data1")[-1].split(":")[0])
        print("dataLength:", dataLength)
        f.close()
        return dataLength

    def _getChannelData(self):
        blockSize = int(self.dataLength / self.nbrBlock)

        f = open(self.hdfFilePath, 'rb')
        f.seek(self.startPoint)
        while True:
            buf = f.read(blockSize)
            if len(buf) == 0:
                break
            self._parseBlock(buf)

        f.close()

    def _parseBlock(self, block):
        index = 0
        for channel in self.channels:
            numReads = int(channel.frequency / self.baseBlockFrequency)
            for i in range(numReads):
                datum = struct.unpack('<f', block[index:index+4])[0]
                channel.addData(datum)
                index = index + 4

File: data_handler/lib/src/NVH_deprecated/hdf_reader.dart
import 'dart:convert';
import 'dart:typed_data';

import './channel_data_model.dart';

class HdfReader {
  int baseBlockFrequency = 1000; // assume 1kHz, should be parsed from delta value
  Uint8List bytes;
  int dataLength = 0;
  int startPoint = 65536; // first candidate
  int numberOfChannel = 0;
  int nbrBlock = 0;
  List<ChannelDataModel> channels = [];

  HdfReader(this.bytes);

  List<ChannelDataModel> parseSync() {
    // parse header
    _parseHeader();

    // find data start point
    dataLength = _getDataLength();

    // parse Raw data to CHannel data model
    _getChannelData();

    return channels;
  }

  _getDataLength() {
    var dataLength = 0;
    var dummy = bytes.sublist(startPoint - 8192, startPoint);
    var dummyString = String.fromCharCodes(dummy);
    if (dummyString.contains("data1")) {
      dataLength = int.parse(dummyString.split("data1").last.split(":").first);
    }
    print("dataLength: $dataLength");
    return dataLength;
  }

  _parseHeader() {
    var headerString = String.fromCharCodes(bytes.sublist(0, startPoint));
    List<String> lines = const LineSplitter().convert(headerString);

    var chOrder = "";
    for (var line in lines) {
      if (line.contains("start of data")) {
        startPoint = int.parse(line.split(":").last);
        print("startPoint: $startPoint");
      }
      if (line.contains("ch order")) {
        chOrder = (line.split(":").last);
        print("Channel Order: $chOrder");
        //TODO -> make channels data map
      }
      if (line.contains("scan mode")) {
        var mode = (line.split(":").last);
        assert(mode != "synchronised multiple");
      }
      if (line.contains("nbr of scans")) {
        nbrBlock = int.parse(line.split(":").last);
        print("Number of Block: $nbrBlock");
        break;
      }
    }

    _parseChannels(headerString.split("distribution func").last, chOrder);
  }

  _parseChannels(String channelDefinitions, chOrder) {

    var chOrders = chOrder.split(",");

    var splits = channelDefinitions.split("implementation type");
    for (var i = 0; i < chOrders.length; ++i) {
      var split = splits[i];
      var chOrder = chOrders[i];

      List<String> lines = const LineSplitter().convert(split);
      var name = "";
      var unit = "";
      var frequency = 0; 
      for (var line in lines) {
        if(line.contains("name str")){
          name = (line.split("name str:").last.replaceAll(" ", ""));
        }
        if(line.contains("physical unit")){
          unit = (line.split(":").last.replaceAll(" ", ""));
        }
      }
      if(name == ""){ break; }

      if(chOrder.contains("*")){
        frequency = int.parse(chOrders.first.split("*").first);
      }
      else{
        frequency = 1;
      }
      frequency *= baseBlockFrequency; // assume block is based on 1kHz.

      channels.add(ChannelDataModel(name, unit, frequency));
      print("name: $name, unit: $unit, frequency: $frequency");
    }
  }

  _getChannelData(){
    var blockSize = dataLength ~/ nbrBlock;

    for(int offset = 0; offset < dataLength; offset += blockSize){
      _parseBlock(bytes.sublist(startPoint+offset, startPoint+offset+blockSize));
    }
  }

  _parseBlock(Uint8List block){
    var index = 0;
    for(var channel in channels){
      var numReads = channel.frequency ~/ baseBlockFrequency;
      for (int i = 0; i < numReads; ++i) {
        var datum = block.buffer.asByteData().getFloat32(index, Endian.little);
        channel.addData(datum);
        index += 4;
      }
    }
  }
}
